ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/PeerSelection.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Ouroboros.Consensus.Shelley.Ledger.PeerSelection () where
+
+import           Data.Bifunctor (second)
+import           Data.Foldable (toList)
+import           Data.List (sortOn)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Ord (Down (..))
+
+import           Ouroboros.Consensus.Ledger.SupportsPeerSelection
+
+import qualified Shelley.Spec.Ledger.Delegation.Certificates as SL
+import qualified Shelley.Spec.Ledger.Keys as SL
+import qualified Shelley.Spec.Ledger.LedgerState as SL
+import qualified Shelley.Spec.Ledger.TxData as SL
+
+import           Ouroboros.Consensus.Shelley.Ledger.Block
+import           Ouroboros.Consensus.Shelley.Ledger.Ledger
+
+instance LedgerSupportsPeerSelection (ShelleyBlock c) where
+  getPeers ShelleyLedgerState { shelleyState } = concat
+      [ Map.findWithDefault [] stakePool poolDomainAddresses
+      | stakePool <- orderByStake poolDistr
+      ]
+    where
+      poolDistr :: SL.PoolDistr c
+      poolDistr = SL.nesPd shelleyState
+
+      -- | Sort stake pools by descending stake
+      orderByStake :: SL.PoolDistr c -> [SL.KeyHash 'SL.StakePool c]
+      orderByStake =
+            map fst
+          . sortOn (Down . snd)
+          . map (second SL.individualPoolStake)
+          . Map.toList
+          . SL.unPoolDistr
+
+      -- | Note that a stake pool can have multiple registered relays
+      poolDomainAddresses :: Map (SL.KeyHash 'SL.StakePool c) [DomainAddress]
+      poolDomainAddresses =
+            Map.map (map relayToDomainAddress . toList . SL._poolRelays)
+          . SL._pParams
```

Comment:
FYI, if we change this from `SL._pParams` to `SL._fPParams`, we get the *future* pool parameters.

Reply:
Presumably that's the set that is updated when pools re-register (that we take snapshots from).

We should aim for:
1. sampling pools from an appropriate stake distribution (e.g. the one used for block production in the current epoch)
2. selecting the pool relays that are the most up to date, so that we can respond to changed relays more quickly than waiting 5 days

Reply:
About 1: I believe the current version of this PR uses that stake distribution. It "samples" it in that it returns *all registered relays* for each stake pool. So the list returned by `getPeers` will look something like (with `pool1` having the most stake, etc.):
```
[pool1_relay1, pool1_relay2, pool2_relay1, pool3_relay1, pool3_relay2, ..]
```

About 2: I suppose we can merge the relays from `SL._pParams` with those of `SL._fPParams` so we get the future relays as well. For example:
```
[pool1_relay1, pool1_relay2, pool1_future_relay3, ..]
```

Let me know what is wanted, I'm just the conduit between network and ledger.

Reply:
I think we need both current and future relays. The governor should try future relays first and if they are not yet accessible then resort to the current ones.

 Another point is that it would be more useful to group relays for the same core node, would that be possible? e.g. `[[pool1_relay1, pool1_relay2, pool1_future_relay3], ...]`. @dcoutts I think this would require some changes to the governor, but it's more useful to know in advance that which relays belong to the same stake pool, instead of relaying on `tip-sample` to discover that through some heuristics.

Reply:
As mentioned, this PR already groups them per stake pool.

@coot Which order do you want for the relays? Future relays first, then the current ones? The other way around? You probably want the duplicates filtered out, right?

Reply:
I think it makes more sense to get future relays first, and we'd like to filter out duplicates.